7538 - Validate cookie when restoring the app from the bfcache

m5r opened the pull request:
# Description

The bug had to do with the browser's bfcache, see [related documentation](https://web.dev/bfcache/)  
Essentially, it restored the page and its state from memory without even hitting the network. Showing a half logged in/half logged out state when coming back to the app from the login page with the browser's back button.

With this PR, when the app is restored from the bfcache, the session is validated once more to make sure the user doesn't get stuck in an impossible state.

medic/cht-core#7538

# Code review checklist
<!-- Remove or comment out any items that do not apply to this PR; in the remaining boxes, replace the [ ] with [x]. -->
- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

dianabarsan reviewed (commented):
Nice work! 
Three things: 
1. Please add an e2e test for this
2. Please add the same logic in the admin app
3. This happened a few times, where I loaded the login page when doing a back. 
The spinner appears immediately (without me clicking anything) and stays there forever, and the button is disabled, just like cache was also used to load the page. 
![image](https://user-images.githubusercontent.com/35681649/161543691-e7f4944f-5e8b-4bd2-8868-0e0905ef08ae.png)

latin-panda commented on tests/wdio.conf.js:
```diff
@@ -78,7 +79,7 @@ const baseConfig = {
     acceptInsecureCerts: true,
     'goog:chromeOptions': {
-      args: ['--headless', '--disable-gpu', '--deny-permission-prompts']
+      args: [/*'--headless',*/ '--disable-gpu', '--deny-permission-prompts']
```
just curious, why are we commenting the `--headless`?

m5r replied:
both changes in `wdio.conf.js` are temporary, I'm trying to reproduce a bug in the CI that I have locally, sorry for the disruption

latin-panda commented on tests/wdio.conf.js:
```diff
@@ -40,7 +40,8 @@ const baseConfig = {
   //
   specs: [
-    './tests/e2e/**/*.wdio-spec.js'
+    // './tests/e2e/**/*.wdio-spec.js',
```
Do we need this commented line?

m5r commented on admin/src/js/services/session.js:
```diff
@@ -114,6 +114,8 @@ const _ = require('lodash/core');
 
         init: checkCurrentSession,
 
+        checkCurrentSession: checkCurrentSession,
```
I've defined another export here named `checkCurrentSession` so it conveys better the intent when calling `Session.checkCurrentSession()`
I couldn't find any `Session.init()` call in the admin app so I wasn't sure if I could simply rename that export

latin-panda replied:
Yes, same, I didn't find any usage of `Session.init()`, I'd remove it and just have `Session.checkCurrentSession()`

m5r commented on admin/src/js/services/db.js:
```diff
@@ -22,6 +22,7 @@ angular.module('inboxServices').factory('DB',
     const isOnlineOnly = Session.isOnlineOnly();
 
     const getUsername = remote => {
+      Session.checkCurrentSession();
```
This gets called immediately when coming back to the admin app with the back button. It strangely gets called before the `pageshow` event handler and this function throwing an error results in the event handler not being called 🤨

But hey, no need for an extra event listener at least :)

m5r commented:
@dianabarsan @latin-panda could you review this PR once more? I've fixed the problem in the admin app, on the login page, and I've added e2e tests for those 3 scenarios

latin-panda commented on tests/wdio.conf.js:
```diff
@@ -40,7 +40,8 @@ const baseConfig = {
   //
   specs: [
-    './tests/e2e/**/*.wdio-spec.js'
+    // './tests/e2e/**/*.wdio-spec.js',
```
Remember to revert this one when you've finish testing :)

m5r replied:
Oops, thanks!

dianabarsan reviewed (commented):
Nice! Just one question inline.

dianabarsan commented on webapp/src/ts/services/session.service.ts:
```diff
@@ -73,6 +78,13 @@ export class SessionService {
       return this.logout();
     }
+
+    window.addEventListener('pageshow', (event) => {
```
I edited the above comment (better code). I was sleepy when I first submitted, sorry.

m5r replied:
Oh, thanks, that's much cleaner indeed! I bet Angular also has some logic to remove event listeners when appropriate.  
And take care of your sleep :zzz:

dianabarsan reviewed (approved):
Nice work!
